File: Java/civitas/src/civitas/TituloPropiedad.java
package civitas;

public class TituloPropiedad {
    private String nombre;
    private float alquilerBase;
    private float factorDeRevalorizacion;
    private float precioHipotecaBase;
    private float precioCompra;
    private float precioEdificar;
    private boolean hipotecado;
    private static float factorInteresesHipoteca = (float) 1.1;
    private int numCasas;
    private int numHoteles;
    private Jugador propietario = null;
    
    
    TituloPropiedad(String _nombre, float _alquilerBase, int _factorDeRevalorizacion, int _precioHipotecaBase, int _precioCompra, int _precioPorEdificar){
        this.propietario = null;
        nombre=_nombre;
        alquilerBase = _alquilerBase;
        factorDeRevalorizacion = _factorDeRevalorizacion;
        precioHipotecaBase = _precioHipotecaBase;
        precioCompra = _precioCompra;        
        precioEdificar = _precioPorEdificar;
        numCasas=0;
        numHoteles=0;
        hipotecado=false;
    }

    public String getNombre() {
        return nombre;
    }

    public float getAlquilerBase() {
        return alquilerBase;
    }

    public float getFactorDeRevalorizacion() {
        return factorDeRevalorizacion;
    }

    public float getPrecioHipotecaBase() {
        return precioHipotecaBase;
    }

    public float getPrecioCompra() {
        return precioCompra;
    }

    public float getPrecioEdificar() {
        return precioEdificar;
    }

    public boolean getHipotecado() {
        return hipotecado;
    }

    public float getFactorInteresesHipoteca() {
        return factorInteresesHipoteca;
    }

    public int getNumCasas() {
        return numCasas;
    }

    public int getNumHoteles() {
        return numHoteles;
    }
    
    public float getPrecioAlquiler(){
        if(hipotecado==true || propietario.isEncarcelado())
            return 0;
        else
            return this.alquilerBase;
    }
    
    public boolean propietarioEncarcelado(){
        return this.propietario.isEncarcelado();
    }
    
    public float getImporteCancelarHipoteca(){
        return this.precioHipotecaBase*factorInteresesHipoteca;
    }
    
    public boolean cancelarHipoteca(Jugador _jugador){
        if(this.esEsteElPropietario(_jugador) && this.hipotecado==true){
            _jugador.paga(this.getImporteCancelarHipoteca());
            this.hipotecado=false;
            return true;
        }
        else
            return false;
        
    }
    
    public boolean esEsteElPropietario(Jugador _jugador){
        return this.propietario==_jugador;
    }
    @Override
    public String toString(){
        String str;
        str= "\n Titulo: {" + "nombre=" + nombre + ", precio de compra= " + Float.toString(precioCompra) + ", alquiler de base= " + Float.toString(alquilerBase) + "factor de revalorización=" + factorDeRevalorizacion + ", precio hipoteca= " + Float.toString(precioHipotecaBase) + ", coste de edificar= " + Float.toString(precioEdificar) + ", numero de casas= "  + Integer.toString(numCasas) + "numero de hoteles=" + Integer.toString(numHoteles) +  "}";
    
        if(hipotecado==false)
            str=str +"Propiedad NO hipotecada }";
        else
            str= str + "Propiedad Hipotecada }";
        
        return str;
    }
            
    void actualizaPropietarioPorConversion(Jugador _jugador){
       propietario = _jugador;
   } 
   
   
   private int cantidadCasasHoteles(){
       return this.numCasas+this.numHoteles;
   }
   
   
   boolean comprar(Jugador _jugador){
       boolean result = false;
       if(this.tienePropietario()==false){
           this.actualizaPropietarioPorConversion(_jugador);
           propietario.paga(this.precioCompra);
           result = true;
       }
       return result;
   }
   
    boolean construirCasa(Jugador _jugador){
       boolean resul=false;
       if(_jugador==this.propietario){
           _jugador.paga(this.getPrecioEdificar());
           this.numCasas++;
           resul=true;
       }
       return resul;
   }
   
    boolean construirHotel(Jugador _jugador){
       boolean resul=false;
       if(_jugador==this.propietario){
           _jugador.paga(this.getPrecioEdificar());
           this.numHoteles++;
           resul=true;
       }
       return resul;
   }
   
    boolean derruirCasa(int n, Jugador _jugador){
       if(_jugador==this.propietario && this.numCasas>=n){
           this.numCasas=this.numCasas-n;
           return true;
       }
       else
           return false;
   }
   
   
   private float getPrecioVenta(){
       return (float) this.precioCompra+((this.precioEdificar*this.getNumCasas()+this.precioEdificar*this.getNumHoteles())*this.factorDeRevalorizacion);
   }
   
   protected Jugador getPropietario(){
       return this.propietario;
   }
   
   protected boolean hipotecar(Jugador _jugador){
       if(this.hipotecado==false && _jugador==this.propietario){
           _jugador.recibe(this.precioHipotecaBase);
           this.hipotecado=true;
           return true;
       }
       else
           return false;
   }
   
    boolean tienePropietario(){
       if(this.propietario==null)
           return false;
       else
           return true;
   }   
   
   protected void tramitarAlquiler(Jugador _jugador){
       if(this.tienePropietario() && _jugador!=this.propietario){
           _jugador.pagaAlquiler(alquilerBase);
           this.propietario.recibe(alquilerBase);
       }
   }
   
   protected boolean vender(Jugador _jugador){
       if(_jugador==propietario){
		_jugador.modificarSaldo(this.getPrecioVenta());
		propietario=null;
		return true;
	}
	else
		return false;
        
   }
   
   
   
}

File: Ruby/civitas/lib/titulo_propiedad.rb
#encoding: utf­8

require_relative 'jugador'

module Civitas
  class TituloPropiedad
    @@factor_intereses_hipoteca=1,1
    
    attr_accessor :hipotecado, :nombre, :num_casas, :num_hoteles, :precio_compra, :precio_edificar, :hipoteca_base, :alquiler_base , :propietario
    
    def initialize(_nombre, _ab, _fr, _hb, _pc, _pe)
      @alquiler_base=_ab
      @factor_revalorizacion=_fr
      @hipoteca_base=_hb
      @hipotecado=false
      @nombre=_nombre
      @num_casas=0
      @num_hoteles=0
      @precio_compra=_pc
      @precio_edificar=_pe
      @propietario = nil
    end
    
    def actualizar_propietario_por_converision(_jugador)
      @propietario=_jugador
    end
    
    def cancelar_hipoteca(_jugador)
      if(self.es_este_el_propietario(_jugador)&& @hipotecado == true)
        _jugador.paga(self.get_importe_cancelar_hipoteca())
        @hipotecado =false
        return true
      else
        return false
      end
      
    end
    
    def cantidad_casahoteles()
      return @num_casas+@num_hoteles
    end
    
    def comprarT(_jugador)
      if(self.tiene_propietario()==true)
        puts "TIENE PROPIETARIO"
        return false
      else
        if(_jugador.paga(@precio_compra))
            actualizar_propietario_por_converision(_jugador)
            return true
        end
      end
    end
    
    def construir_casa(_jugador)
      if(_jugador == @propietario)
        _jugador.paga(@precio_edificar)
        @num_casas = @num_casas + 1
        return true
      else
        return false
      end
    end
    
    def construir_hotel(_jugador)
      if(_jugador == @propietario)
        _jugador.paga(@precio_edificar)
        @num_hoteles = @num_hoteles + 1
        return true
      else
        return false
      end
    end
    
    def derruir_casa(_n, _jugador)
      if(_jugador == @propietario and @num_casas >= _n.to_f)
        @num_casas = 0
        return true
      else
        return false
      end
    end
    
    def es_este_el_propietario(_jugador)
      return @propietario== _jugador
    end
    
    def get_importe_cancelar_hipoteca()
      cantidad = @hipoteca_base*(@num_casas + @num_hoteles)
      return cantidad
    end
    
    def get_precio_venta()
      return @precio_compra + ((@precio_edificar * @num_casas + @precio_edificar * @num_hoteles)*@factor_revalorizacion)
    end
    
    
    
    def hipotecarT(_jugador)
      puts "HIPOTECART"
      puts @propietario.inspect
      if(@hipotecado == false && _jugador == @propietario)
        puts "IF HIPOTECART"
        _jugador.recibe(@hipoteca_base)
        @hipotecado = true
        return true
      else
        return false
      end
    end
    
    def propietario_encarcelado()
      return @propietario.encarcelado
    end
    
    def tiene_propietario()
      if(@propietario == nil)
        return false
      else
        return true
      end
    end
    
    def get_num_casas()
      return @num_casas
    end
    
    def get_num_hoteles()
      return @num_hoteles
    end
    
     @Override
     def to_st()
       final = "Titulo: Nombre #{@nombre} Compra #{@precio_compra.to_s()} alquilerBase #{@alquier_base.to_s()} FactorRevalorizacion #{@factor_revalorizacion.to_s()} hipotecaBase #{@hipoteca_base.to_s()} precioCompra #{@precio_compra.to_s()} precioEdificar #{@precio_edificar.to_s()} numero de Casas #{@num_casas.to_s()} numero de Hoteles #{@num_hoteles.to_s() }"
       
       if(@hipotecado == false)
         final += "Propiedad no hipotecada"
       else
         final += "Propiedad hipotecada"
       end
     end
    
     
     def tramitar_alquiler(_jugador)
       if(self.tiene_propietario() && _jugador!=@propietario)
         _jugador.paga_alquiler(@alquier_base)
         @propietario.recibe(@alquiler_base)
       end
     end
     
     def venderT(_jugador)
       if(@propietario==_jugador)
         puts "IF DE VENDERT"
         _jugador.modificar_saldo(get_precio_venta())
         @propietario=nil
         return true
       else
         return false
       end
     end
      
  end

end
